bridges/ethMultiversX/bridgeExecutor.go

```diff
@@ -575,69 +575,128 @@ func (executor *bridgeExecutor) checkCumulatedTransfers(ctx context.Context, tok
+	isNativeOnMultiversX := executor.isNativeOnMultiversX(ctx, convertedToken)
+
+	if isNativeOnEthereum && isNativeOnMultiversX {
+		return fmt.Errorf("%w isNativeOnEthereum = %v, isNativeOnMultiversX = %v", ErrInvalidSetup, isNativeOnEthereum, isNativeOnMultiversX)
 	}
-	return isMintBurnOnEthereum, nil
-}
 
-func (executor *bridgeExecutor) checkRequiredMintBurnBalance(ctx context.Context, ethToken common.Address, mvxToken []byte) error {
-	ethBalance, err := executor.ethereumClient.TokenMintedBalances(ctx, ethToken)
-	if err != nil {
-		return err
+	if !isMintBurnOnEthereum && !isMintBurnOnMultiversX {
+		return fmt.Errorf("%w isMintBurnOnEthereum = %v, isMintBurnOnMultiversX = %v", ErrInvalidSetup, isMintBurnOnEthereum, isMintBurnOnMultiversX)
 	}
 
-	mvxBalance, err := executor.multiversXClient.AccumulatedBurnedTokens(ctx, mvxToken)
+	ethAmount, err := executor.computeEthAmount(ctx, token, isMintBurnOnEthereum, isNativeOnEthereum)
 	if err != nil {
 		return err
 	}
-	if ethBalance.Cmp(mvxBalance) > 0 {
-		return fmt.Errorf("%w, balance for ERC20 token %s is %s and the balance for ESDT token %s is %s",
-			ErrMintBurnBalance, ethToken.String(), ethBalance.String(), mvxToken, mvxBalance.String())
+	mvxAmount, err := executor.computeMvxAmount(ctx, convertedToken, isMintBurnOnMultiversX, isNativeOnMultiversX)
+	if err != nil {
+		return err
 	}
 
 	executor.log.Debug("bridgeExecutor.checkRequiredMintBurnBalance",
-		"ERC20 token", ethToken.String(),
-		"ERC20 native balance", ethBalance.String(),
-		"ESDT token", mvxToken,
-		"ESDT native balance", mvxBalance.String(),
+		"ERC20 token", token.String(),
+		"ERC20 native balance", ethAmount.String(),
+		"ESDT token", convertedToken,
+		"ESDT native balance", mvxAmount.String(),
 	)
 
+	mvxAmount = big.NewInt(0).Sub(mvxAmount, amount)
+	if ethAmount.Cmp(mvxAmount) != 0 {
+		return fmt.Errorf("%w, balance for ERC20 token %s is %s and the balance for ESDT token %s is %s",
+			ErrBalanceMismatch, token.String(), ethAmount.String(), convertedToken, mvxAmount.String())
+	}
 	return nil
 }
 
+func (executor *bridgeExecutor) checkRequiredBalance(ctx context.Context, token common.Address, convertedToken []byte, amount *big.Int) error {
+	// transfer to Ethereum
+	if amount.Cmp(big.NewInt(0)) < 0 {
+		return executor.ethereumClient.CheckRequiredBalance(ctx, token, amount)
+	}
+	// transfer to MultiversX
+	return executor.multiversXClient.CheckRequiredBalance(ctx, convertedToken, amount)
+}
+
+func (executor *bridgeExecutor) computeEthAmount(ctx context.Context, token common.Address, isMintBurn bool, isNative bool) (*big.Int, error) {
+	if !isMintBurn {
+		return executor.ethereumClient.TotalBalances(ctx, token)
+	}
+
+	burnBalances, err := executor.ethereumClient.BurnBalances(ctx, token)
+	if err != nil {
+		return big.NewInt(0), err
+	}
+	mintBalances, err := executor.ethereumClient.MintBalances(ctx, token)
+	if err != nil {
+		return big.NewInt(0), err
+	}
+
+	if isNative {
```

Comment:
should we also check if burnBalances >= mintBalances? similar for L64, L662, L664

Reply:
added